File: 0649-dota2-senate/0649-dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        rQueue = deque()
        dQueue = deque()
        n = len(senate)

        for i, x in enumerate(senate):
            if x == 'R': rQueue.append(i)
            else: dQueue.append(i)
        

        while rQueue and dQueue:
            if rQueue[0] < dQueue[0]:
                dQueue.popleft()
                rQueue.popleft()
                rQueue.append(n + 1)
            else:
                rQueue.popleft()
                dQueue.popleft()
                dQueue.append(n + 1)
            n += 1
        
        if rQueue:
            return 'Radiant'
        
        return 'Dire'

File: 0649-dota2-senate/0649-dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> r, d;
        for (int i = 0; i < senate.size(); i++) {
            if (senate[i] == 'R') {
                r.push(i);
            } else {
                d.push(i);
            }
        }

        int i = senate.size();
        while (r.size() && d.size()) {
            if (r.front() < d.front()) {
                d.pop();
                r.pop();
                r.push(i);
            } else {
                r.pop();
                d.pop();
                d.push(i);
            }

            i++;
        }

        if (r.size()) {
            return "Radiant";
        } else {
            return "Dire";
        }
    }
};